### PR TITLE
fix: cannot get embedding from reloaded DocumentArrayMemmap

### PR DESCRIPTION
**Pull request by winstonww**
This PR aims to close issue #2944 . Thank you.

### CODE REVIEW

**Review by JoanFM — COMMENTED** [review]
Let's add a test that shows the motivation for the change

**Comment by winstonww** [comment]
@JoanFM Sure, thanks for the suggestion.

**Review by JoanFM — CHANGES_REQUESTED** [review]
remember to add test

**Review by numb3r3 — CHANGES_REQUESTED** [review]
This PR needs improvement. The following test cannot pass:
```
from jina import Document
from jina.types.arrays.memmap import DocumentArrayMemmap
import numpy as np

def random_docs(num_docs):
    for j in range(num_docs):
        doc_id = j

        d = Document(id=doc_id)
        d.tags['id'] = doc_id
        
        d.embedding = np.random.random(
            [10 + np.random.randint(0, 1)]
        )
        yield d

dam = DocumentArrayMemmap('tmp_memmap_toy')
docs = list(random_docs(100))
dam.extend(docs)

dam2 = DocumentArrayMemmap('tmp_memmap_toy')
x = dam2.get_attributes('embedding')
assert len(dam2) == 100


docs = list(random_docs(100))
dam2.extend(docs)
x = dam2.get_attributes('embedding')
assert len(dam2) == 200
```
BTW., please convert the above script to a pytest

**Comment by winstonww** [comment]
Thank you for the suggestion. I think the error happens because in the following:
>```
> docs = list(random_docs(100))
> dam2.extend(docs)
> x = dam2.get_attributes('embedding')
> assert len(dam2) == 200
> ```

docs generated by this `random_docs` has the same id as the docs generated by the previous `random_docs`. As a result, docs are getting replaced instead of extended. 

If we replace `random_docs`  implementation with the following:
```
def random_docs(num_docs, start=0):
    for j in range(start, num_docs+start):
       ...
```
Then the proposed case passes.  I will continue to see if there's any other issues with the PR in the meantime.

**Comment by numb3r3** [comment]
> Thank you for the suggestion. I think the error happens because in the following:
> 
> > ```
> > docs = list(random_docs(100))
> > dam2.extend(docs)
> > x = dam2.get_attributes('embedding')
> > assert len(dam2) == 200
> > ```
> 
> docs generated by this `random_docs` has the same id as the docs generated by the previous `random_docs`. As a result, docs are getting replaced instead of extended.
> 
> If we replace `random_docs` implementation with the following:
> 
> ```
> def random_docs(num_docs, start=0):
>     for j in range(start, num_docs+start):
>        ...
> ```
> 
> Then the proposed case passes. I will add the above to pytest and continue to see if there's any other issues with the PR in the meantime.

I see, the name `DocumentArrayMemmap` is very confusing. This is named as an array, but this is actually a map-style data.

File: tests/unit/types/arrays/test_memmap.py [hunk 1]
```diff
@@ -173,3 +173,19 @@ def test_convert_dm_to_dam(tmpdir, mocker):
...
+    dam.extend(docs)
+
+    dam2 = DocumentArrayMemmap(tmpdir)
+    x = dam2.get_attributes('embedding')
+    assert len(dam2) == 100
+
+    docs = list(random_docs(100, start_id=100, embed_dim=embed_dim))
+    dam2.extend(docs)
+    x = dam2.get_attributes('embedding')
+    assert len(dam2) == 200
```
Comment (numb3r3) [inline tests/unit/types/arrays/test_memmap.py:191]:
```suggestion
    dam2 = DocumentArrayMemmap(tmpdir)
    x = dam2.get_attributes('embedding')
    assert len(dam2) == 100
    assert len(x) == 100
    assert x[0].shape == (embed_dim,)

    docs = list(random_docs(100, start_id=100, embed_dim=embed_dim))
    dam2.extend(docs)
    x = dam2.get_attributes('embedding')
    assert len(dam2) == 200
    assert len(x) == 200
    assert x[0].shape == (embed_dim,)
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: tests/unit/types/arrays/test_memmap.py (hunk 1 above)
Comment (winstonww) [reply tests/unit/types/arrays/test_memmap.py]: Addressed, please see [e5dea0f](https://github.com/jina-ai/jina/pull/2956/commits/e5dea0f45fbd88593f8ae29405a8c43085a99996) for the change, thanks.

**Review by numb3r3 — APPROVED** [review]
LGTM

**Comment by florian-hoenicke** [comment]
I would recommend:
Whenever there is a corrupted DocumentArray on the disk, Jina should log a warning and initialize an empty DocumentArray.

File: tests/unit/types/arrays/test_memmap.py [hunk 2]
```diff
@@ -173,3 +173,22 @@ def test_convert_dm_to_dam(tmpdir, mocker):
     mock.assert_called()
     assert len(da) == 0
     assert len(dam) == 100
+
```
Comment (florian-hoenicke) [inline tests/unit/types/arrays/test_memmap.py:176]: Can you test if manual interruption of the writing process causes corruptions and therefore this error message?

**Review by florian-hoenicke — CHANGES_REQUESTED** [review]

**Comment by florian-hoenicke** [comment]
Issue @alexcg1 discovered:
```
Clone https://github.com/alexcg1/jina-shortest-program
Run python sensible.py
Ctrl-C
Don't delete workspace
Re-run the script
Try to query via curl
```

File: jina/types/arrays/memmap.py [hunk 3]
```diff
@@ -147,7 +147,7 @@ def append(self, doc: 'Document', flush: bool = True) -> None:
             ).tobytes()
         )
         self._header_map[doc.id] = (len(self._header_map), p, r, r + l)
-        self._start = p + r + l
+        self._start = p + r + l  # <==> self._start = self._start + l
```
Comment (hanxiao) [inline jina/types/arrays/memmap.py:150]: can't understand the comment here, suggestion: either delete it or write it clearly

```suggestion
        self._start = p + r + l
```

**Review by hanxiao — CHANGES_REQUESTED** [review]
see comment above

**Comment by hanxiao** [comment]
#2944 looks like a fake issue. hold the merge

**Pull request closed without merging**